File: 并查集/省份数量.java
package 并查集;

public class 省份数量 {
    // 节点i属于集群father[i]
    int[] father;

    public int findCircleNum(int[][] isConnected) {
        int n = isConnected.length;
        father = new int[n];
        init();
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                if (isConnected[i][j] == 1) join(i, j);
            }
        }
        int count = 0;
        for (int i = 0; i < n; i++) {
            if (father[i] == i) {
                count++;
            }
        }
        return count;
    }

    public void init() {
        for (int i = 0; i < father.length; i++) {
            father[i] = i;
        }
    }

    public int find(int x) {
        if (x == father[x]) return x;
        father[x] = find(father[x]);
        return father[x];
    }

    public void join(int x, int y) {
        x = find(x);
        y = find(y);
        if (x == y) return;
        father[y] = x;
    }

}

File: 图/并查集/省份数量.go
package 并查集

//计算连通分量的个数
func findCircleNum(isConnected [][]int) int {
	n := len(isConnected) //一共有n个城市
	var father []int = make([]int, n)
	//初始化并查集
	for i := 0; i < len(father); i++ {
		father[i] = i
	}
	var find func(x int) int //找跟节点
	find = func(x int) int {
		if x == father[x] {
			return x
		}
		father[x] = find(father[x])
		return father[x]
	}
	//var same func(x, y int) bool
	//same = func(x, y int) bool {
	//	x = find(x)
	//	y = find(y)
	//	return x == y
	//}
	var join func(x, y int) //将y所在集群加入x所在的集群
	join = func(x, y int) {
		x = find(x)
		y = find(y)
		if x == y {
			return
		}
		father[y] = x
	}
	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			if isConnected[i][j] == 1 {
				join(i, j)
			}
		}
	}
	var numOfCluster int
	for i := 0; i < n; i++ {
		//跟节点的唯一特征就是father[i] == i
		if father[i] == i {
			numOfCluster++
		}
	}
	return numOfCluster
}
